docs/failure-detector/failure-detector.md

```diff
@@ -49,4 +49,76 @@ In the example:
+in the layout and also collects their node states to provide cluster state.
+To provide more resilient failre detection mechanism, the failure detector collects few "poll reports" 
+equal to `failureThreshold` number then aggregates final poll report
+
+Once poll report collected, the `failureDetectionTask` is triggered by Remote Monitoring Service.
+To detect a failure in the cluster the Remote Monitoring Service gets `cluster state` from a poll report.
+The cluster state contains information of the cluster.
+
+**ClusterAdvisor** detects failures using cluster state. There could be different types of cluster advisors.
+Currently we use `CompleteGraphAdvisor` - any alive node must be connected to all other alive nodes in the cluster.
+`CompleteGraphAdvisor`:
+ - The advisor transforms cluster state into the cluster graph and makes it symmetric.
+ It means converting a cluster graph which could have asymmetric failures to a graph with symmetric failures between nodes.
+ For instance: <br/>
+ {a: [{a: OK, b: FAILED}]} <br/>
+ {b: [{a: OK, b: OK}]} <br/>
+ Node A believes that node B is disconnected <br/>
+ Node B believes that node A is connected <br/>
+ The graph will be changed to: <br/>
+ {a: [{a: OK, b: FAILED}]} <br/>
+ {b: [{a: FAILED, b: OK}]} <br/>
+ Node B is not connected to node A anymore.
+ 
+**Decision maker** - is a node which can decide to add a failed node into the unresponsive list.
+The decision maker is an technique, used to reduce parallel updates of the cluster state.
+Still, each node can choose its own decision maker.
+Note: it's not required to choose a particular one for entire cluster to add a node to unresponsive list.
+
+The decision maker must have:
+- highest number of successful connections in the graph.
+There could be many nodes with same number of successful connections, 
+then the decision maker will be a node with smallest name.
+
+We also have additional checks to prevent all possible incorrect ways. Decision maker can't be found if:
+- `ClusterGraph` is empty, which is an invalid state.
+- the decision maker doesnt have connections, which is also impossible.
+
+Detecting **failed node** - find a node in the cluster with a minimum number of successful connections.
+To find failed nodes and heal nodes in the cluster we use `NodeRank`-s.
+
+`NodeRank` is the rank of the node in a graph. Nodes sorted according to their ranks. 
+Sorting nodes according to:
+- Descending order of number of connections
+- Ascending order of a node name (alphabetically)
+
+A `Decision maker` always has highest number of successful connections and smallest name and
+a failed node always has lowest number of successful connections and highest name.
+
+Finding a failed node - collect all node ranks in a graph and choose the node 
+with the smallest number of successful connections and with the highest name (sorted alphabetically).
+
+Even if we have found a node with smallest number of connections it doesnt mean that the node is 
```

Comment:
What is the meaning of fully connected here? Doesn't fully connected means connect to every other nodes?

Reply:
means - the node has connected to all responsive nodes in the cluster

Reply:
I think we should specifically state here that 'fully connected' in this context means fully connected to the subset of alive nodes

Reply:
it is already explained in this docment. See https://github.com/CorfuDB/CorfuDB/blob/master/docs/failure-detector/failure-detector.md
![image](https://user-images.githubusercontent.com/314292/58298370-173ac180-7d90-11e9-983f-dfdbfa83e940.png)